### PR TITLE
Improve Redis failure modes

### PR DESCRIPTION
**Pull request by agnivade**
While investigating code paths during
Redis downtime, I found out that the library
by default retries indefinitely for read commands.

This leads to it being stuck in a loop
while trying to read anything from Redis.
We fix that by disabling retry.

Secondly, we adjust the ConnWriteTimeout
value slightly from its default of 10s
which is a bit too high for normal working
conditions.

Ref: https://github.com/redis/rueidis/issues/680
```release-note
NONE
```

### CODE REVIEW

**Comment by streamer45** [comment]
What are your thoughts on following the upstream advice (i.e., using `RetryDelayFn`)? I still think some degree of retry makes sense in a network-dependentt component.

**Comment by agnivade** [comment]
Agree. That needs a library upgrade, so I was thinking of doing it later, but let me do it in this PR.

**Comment by agnivade** [comment]
Done. PTAL.

**Comment by agnivade** [comment]
So a problem with retrying is that for a single API request, there are around 30-40 cache calls. So if Redis goes down, a request will take `30 * timeout * retry` seconds for it to fail. With a Dial Timeout of 5s and no retry, createPost takes around 2m19s to return when Redis goes down. Retrying even once will likely double that time.

I am slightly leaning towards not retrying to fail as early as possible, rather than giving the indication that the request is stuck.

LMK what you think.

**Comment by streamer45** [comment]
> So a problem with retrying is that for a single API request, there are around 30-40 cache calls. So if Redis goes down, a request will take `30 * timeout * retry` seconds for it to fail. With a Dial Timeout of 5s and no retry, createPost takes around 2m19s to return when Redis goes down. Retrying even once will likely double that time.
> 

Are all those calls run serially?

> I am slightly leaning towards not retrying to fail as early as possible, rather than giving the indication that the request is stuck.
> 
> LMK what you think.

Maybe we could implement a simple "backoff" algorithm, which essentially means tracking retries and failing other calls immediately if we get above a certain threshold. I agree it may not make sense to queue up all these calls if we have confidence they will fail (e.g. connection is gone).

File: server/platform/services/cache/provider.go [hunk 1]
```diff
@@ -100,6 +101,17 @@ func NewRedisProvider(opts *RedisOptions) (Provider, error) {
 		// meant to be used at very high scales. The docs suggest 20us,
 		// but going as high as 250us doesn't make any material difference.
 		MaxFlushDelay: 250 * time.Microsecond,
+		RetryDelay: func(attempts int, _ rueidis.Completed, _ error) time.Duration {
+			// We only retry once.
+			if attempts > 1 {
+				return -1
+			}
```
Comment (streamer45) [inline server/platform/services/cache/provider.go:108]: Does the `attempts` counter increase per-call or globally (asking since the client is shared)?

**Comment by streamer45** [comment]
> I am slightly leaning towards not retrying to fail as early as possible, rather than giving the indication that the request is stuck.

Yeah, given the presence of a fallback that may be best indeed. Ideally I'd like to track the state of the Redis connection separately (e.g., through a ping or something) and avoid the per-request complication and fallback immediately. But that will require more work.

File: server/platform/services/cache/provider.go (hunk 1 above)
Comment (agnivade) [reply server/platform/services/cache/provider.go:108]: It is per call.

**Comment by agnivade** [comment]
> and avoid the per-request complication and fallback immediately. But that will require more work.

Correct. It requires wrapping the redis client with another struct and fail immediately if one call has failed. A bit of additional work.

**Review by streamer45 — APPROVED** [review]
Appreciate the apparent back and forth discussion. If anything it helped me to understand this much better.

**Pull request merged**